src/java.base/share/classes/sun/security/jca/ProviderConfig.java

```diff
@@ -176,60 +176,63 @@ Provider getProvider() {
+                        new PrivilegedAction<Provider>() {
+                            public Provider run() {
+                                try {
+                                    Class<?> c = Class.forName(
+                                        "apple.security.AppleProvider");
+                                    if (Provider.class.isAssignableFrom(c)) {
+                                        @SuppressWarnings("deprecation")
+                                        Object tmp = c.newInstance();
+                                        return (Provider) tmp;
+                                    }
+                                } catch (Exception ex) {
+                                    if (debug != null) {
+                                        debug.println("Error loading provider Apple");
+                                        ex.printStackTrace();
+                                    }
+                                }
                                 return null;
                             }
-                        } catch (Exception ex) {
-                            if (debug != null) {
-                                debug.println("Error loading provider Apple");
-                                ex.printStackTrace();
-                            }
-                            return null;
+                        });
+                    yield tmp;
+                }
+                default -> {
+                    if (isLoading) {
+                        // because this method is synchronized, this can only
+                        // happen if there is recursion.
+                        if (debug != null) {
+                            debug.println("Recursion loading provider: " + this);
+                            new Exception("Call trace").printStackTrace();
                         }
+                        yield null;
                     }
-                });
-                p = tmp;
-            } else {
-                if (isLoading) {
-                    // because this method is synchronized, this can only
-                    // happen if there is recursion.
-                    if (debug != null) {
-                        debug.println("Recursion loading provider: " + this);
-                        new Exception("Call trace").printStackTrace();
+                    try {
+                        isLoading = true;
+                        tries++;
+                        yield doLoadProvider();
+                    } finally {
+                        isLoading = false;
                     }
-                    return null;
                 }
-                try {
-                    isLoading = true;
-                    tries++;
-                    p = doLoadProvider();
-                } finally {
-                    isLoading = false;
-                }
-            }
+            };
             provider = p;
         }
-        return p;
+        return provider;
```

Comment:
Why changing it to return 'provider', is this change intentional? IIRC, the general pattern is to return the local reference instead of a volatile variable.

Reply:
Ok